### DISCUSSION TITLE
Enforcing maximum name (identifier) length

### DISCUSSION THREAD

**Discussion by ipspace** [ipspace/netlab#756] [Features]
I added new "identifier" data type -- an alphanumeric string up to 16 characters long -- and use it to validate all sorts of names (nodes, VLANs, VRFs, address pools...). This approach nicely solves a few gotchas like invalid node names (that might trip some network devices) or using numbers as VLAN names.

One of the existing test topologies broke because it used very long VLAN names, so I fixed that, but it made me wonder: is there a good reason to support longer names, and if so, what would a sensible maximum length be?

Any feedback @ssasso @jbemmel @ddutt @petercrocker?

FYI: the VRF names were already limited to 16 characters because they're used as Linux device names. I'm pretty positive there might be a few network devices out there that don't support infinitely-long hostnames or VLAN names either ;)

**Comment by petercrocker** [comment]
@ipspace does the hostname include the domain name part? I assume there's not auto-parsing of a FQDN in to hostname / domain. This would lead me to want 64 characters for the limit.

**Reply by ipspace** [reply]
You can't use FQDNs as node names as that would be interpreted as dot notation and the FQDN split into a hierarchy of dictionaries.

You COULD be really persistent and make it work (probably) with a list of nodes, but why bother... unless you have a use case where you'd love to use FQDNs

**Comment by jbemmel** [comment]
I don't see any good reason to use longer names for lab testing topologies, to me "all identifiers should start with a-z and are limited to at most 16 characters" is a decent general approach.

We may want to make them case insensitive, and normalize to all lower case for example

**Comment by ssasso** [comment]
ok for me as well ;)

**Comment by ipspace** [comment]
Node names can already be longer (configurable default), and it's pretty easy to add the same functionality to other objects that might need longer names.

**Comment by ghost** [comment]
Maybe I'm using netlab incorrectly, but I'm trying to build a one to one hardware out of the loop test environment that's representative of my entire LAN, inclusive of hostnames, and I'm running into a road block here; the 16 character limit for node names, specifically links. I saw and am using MAX_NODE_ID_LENGTH, but that doesn't seem to apply to link definitions.

I'm in the unfortunate position of having some switches that have names as long as 19 characters and can't seem to use the full name. Beyond getting clever with substrings is there any way to have a link definition be longer than 16 characters?

**Reply by ipspace** [reply]
We had to enforce a limit on object name lengths due to various restrictions of the underlying operating system/software objects. The most restrictive one (IIRC) is the 16-character limit of Linux interface names (sometimes we're using Linux bridges to connect stuff), and it was easiest to limit everything to that length, we the option of extending some stuff if needed.

Having said that, it's not too hard to add other nerd knobs like MAX_NODE_ID_LENGTH for other objects (as in "giving you enough rope"). Could you please post a sample topology so I can see where exactly you're hitting the limits?

**Reply by ghost** [reply]
Having that knob would be helpful, having everything match up directly makes it easier grok.


Obfuscated representative sample here.

```
defaults.const.MAX_NODE_ID_LENGTH: 32
defaults.device: eos
provider: clab

nodes: [ ab-bc-cd-efgh-ij,ab-bc-cd-efgh-ik ]

links:
- ab-bc-cd-efgh-ij:
     ifindex: 1
  ab-bc-cd-efgh-ik:
     ifindex: 1
```

**Reply by ipspace** [reply]
Fixed. The **links.interfaces[N].node** attribute should be a node ID, not a generic ID.

To test the fix, please clone the GitHub repo and run netlab from there (https://netlab.tools/install/clone/), and if you encounter any further issues, please let me know. I'll do my best to fix them ASAP.

**Comment by ipspace** [comment]
Looks like we solved this one
